### ISSUE TITLE
qubes-issues bot for warning and closing stale issues

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#7270]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

When a Qubes release reaches EOL, there are typically many issues on that release's milestone still open.

- Leaving the issues open on that milestone doesn't make sense. EOL releases no longer receive support, so it's unclear what it means to haven issue open on such a release's milestone.
- Closing all of the issues is wrong, since many of them probably still merit action.
- Reassigning all the issues to a supported release milestone is wrong, because it's unlikely that all of the issues affect newer releases.
- Going through every issue to figure out whether it affects newer releases isn't feasible.
  - We can't do it ourselves, because it often requires more info from the issue reporter.
  - Issue reporters often fail to respond.
  - Even if it were within our power to figure it all out ourselves, we wouldn't have time to investigate every single issue unless we halted all other project activities (and probably not even then).

### The solution you'd like

Create a bot that does the following each day:
1. Check all open issues on closed milestones. (When a release reaches EOL, we close its milestone.)
2. If there has been no activity in at least one year, automatically post a comment explaining the situation and warning that the issue will be closed if there is no activity in one month.
3. If such a warning was posted at least a month ago, close the issue.

Feel free to replace the values "one year" and "one month" with better ones.

### The value to a user, and who that user might be

Everyone who uses the issue tracker would benefit from there being fewer stale issues to sort through.

In the past, I've tried closing very old issues on EOL milestones with a message explaining the situation and saying that we'd be happy to reopen the issue if anyone is still affected. This freaked a lot of people out. It seems that many people view closing an issue as some kind of permanent action, when in reality it's more like a label that can easily be changed with the press of a single button. Indeed, closed issues are just as searchable as open issues. Nonetheless, we might as well acknowledge that there's some kind of widespread psychological quirk about "closing" issues. The proposed bot solution avoids this problem by warning well in advanced that the issue will be closed. This allows anyone who wishes to keep the issue open to add a comment (or perform any other action on the issue). With the values proposed above, even someone who only checks GitHub _once a year_ would never be caught by surprise.

-----

CC: @marmarek, @fepitre

**Comment by brendanhoar** [comment]
> If such a warning was posted at least a month ago, close the issue.

Would this be regardless of new activity after the posting? 

If so, would there be an alternate method of holding off auto-bot-closure (perhaps driven by qubes team tagging)?

Or would the qubes team just manually re-open the issue if determined to still be valid?

[Asking mostly for process clarification, as I don't really have a strong opinion. Process would likely drive content of bot-posted message, though.]

B

**Comment by andrewdavidwong** [comment]
> > If such a warning was posted at least a month ago, close the issue.
> 
> Would this be regardless of new activity after the posting? 
> 

No. I neglected to clarify that in the quoted portion, but it's made clear by the rest of the issue description. I'll edit the quoted portion now.

**Comment by adrelanos** [comment]
Scary idea...

Usually stale bots are conceived as depressing by me. Closing issues which are stale but unresolved at other projects and then not re-opened.

In many tickets such as random late example https://github.com/QubesOS/qubes-issues/issues/7267 I don't understand why milestone `Release 4.1 updates` was added. Well, in that very example it will likely soon be resolved since I already a pull request was submitted but I think otherwise it would be unlikely that this issue would ever solve itself through a Qubes release upgrade to 4.2 or even 5.0.

Another example is https://github.com/QubesOS/qubes-issues/issues/5667 - happened during Qubes 4.0. Does it still happen with Qubes R4.1? I don't know. Didn't test yet since that would take time. From my perspective, could even be wasted time. Since there was no reply last time, there might be no reply if I report "R4.1 still affected". But if I don't test and comment, the stale bot would close the ticket.

**Comment by DemiMarie** [comment]
What if the bot only marked issues with a label, instead of closing them?

#7267 was given `Release 4.1 updates` because that that is the earliest impacted release.  It could not impact R4.0, if I understand correctly, because the relevant feature did not exist in R4.0.  #5667 should have been investigated; I am not sure why it was not.

**Comment by marmarek** [comment]
I know some projects use "stale" bot in quite aggressive way (like, closing any issue after a month of inactivity). We don't want to do that. We want it to handle only issues that are affecting _no longer supported releases_ (for example, if an issue affects R3.2, it should be closed now, unless it also affects a supported release like R4.0 or R4.1). 
It may happen that some such issues would be closed even if still valid - just because nobody noticed (and commented) it is still valid. But then:
- if nobody complain about an issue, then perhaps it isn't important issue after all, and we should focus on things that are actually annoying to the users?
- if it is still valid, and somebody confirms it (just a bit too late), we can trivially reopen it, that's just one click

**Comment by andrewdavidwong** [comment]
@adrelanos:

> Scary idea...
> 
> Usually stale bots are conceived as depressing by me. Closing issues which are stale but unresolved at other projects and then not re-opened.
> 

But did you read the issue description? I specifically explained why and how that would *not* happen here.

> In many tickets such as random late example #7267 I don't understand why milestone `Release 4.1 updates` was added. 

What do you mean? You reported that bug three days ago, and you didn't specify which Qubes release you discovered the bug in, so I presumed it was in Qubes 4.1. Was it not?

> Well, in that very example it will likely soon be resolved since I already a pull request was submitted but I think otherwise it would be unlikely that this issue would ever solve itself through a Qubes release upgrade to 4.2 or even 5.0.
> 

But that's not how bug reports or milestones work. Every bug is a bug *in* some software that already exists, and the milestone of a bug report is the earliest affected release. You can't report a bug today against a future software release that doesn't exist yet. (Well, technically you could, if you plan to introduce the bug in that future release, but of course no one does that.)

It's fine if you don't understand how the issue tracking system works, but that's why I've documented how it works. Please read the documentation to understand:

https://www.qubes-os.org/doc/issue-tracking/#type
https://www.qubes-os.org/doc/issue-tracking/#milestones

If the documentation is unclear, or you disagree with how things are done and have suggestions for improvement, I'm open to talking about it, but that would be a separate topic of discussion.

> Another example is #5667 - happened during Qubes 4.0. Does it still happen with Qubes R4.1? I don't know. Didn't test yet since that would take time. From my perspective, could even be wasted time. Since there was no reply last time, there might be no reply if I report "R4.1 still affected". But if I don't test and comment, the stale bot would close the ticket.

No, no one said you have to test *and* comment. Just a comment is enough, like, "Still need to check on this," "Might affect 4.1," or even just "Please keep open." The bot doesn't know the difference. It just checks for *any activity whatsoever*.

-----

@DemiMarie:

> What if the bot only marked issues with a label, instead of closing them?

What would be the point of that? In other words, what (if anything) would subsequently be done with issues with that label?

-----

@marmarek:

> It may happen that some such issues would be closed even if still valid - just because nobody noticed (and commented) it is still valid. But then:
> 
> * if nobody complain about an issue, then perhaps it isn't important issue after all, and we should focus on things that are actually annoying to the users?
> * if it is still valid, and somebody confirms it (just a bit too late), we can trivially reopen it, that's just one click

Exactly. This happens even now (without the bot) with regressions. Someone comments on the old (closed) issue, so we reopen it and update the milestone to the newer affected release. This is why our policy is to comment on existing issues (even closed ones) rather than opening duplicates.

**Comment by adrelanos** [comment]
Sounds a lot better to me know.

> @adrelanos:
> 
> > Scary idea...
> > Usually stale bots are conceived as depressing by me. Closing issues which are stale but unresolved at other projects and then not re-opened.
> 
> But did you read the issue description? I specifically explained why and how that would _not_ happen here.

Expected implementation versus actual implementation. The vision might be great but I imagined how it could be actually implemented. The stale bot might close too many issues and when this gets reported as a bug it might be closed as "not our bug".

> > In many tickets such as random late example #7267 I don't understand why milestone `Release 4.1 updates` was added.
> 
> What do you mean?

The explanation provided below was helpful.

> You reported that bug three days ago, and you didn't specify which Qubes release you discovered the bug in, so I presumed it was in Qubes 4.1. Was it not?

That was correct.

> > Well, in that very example it will likely soon be resolved since I already a pull request was submitted but I think otherwise it would be unlikely that this issue would ever solve itself through a Qubes release upgrade to 4.2 or even 5.0.
> 
> But that's not how bug reports or milestones work. Every bug is a bug _in_ some software that already exists, and the milestone of a bug report is the earliest affected release. You can't report a bug today against a future software release that doesn't exist yet. (Well, technically you could, if you plan to introduce the bug in that future release, but of course no one does that.)

> It's fine if you don't understand how the issue tracking system works, but that's why I've documented how it works. Please read the documentation to understand:
> 
> https://www.qubes-os.org/doc/issue-tracking/#type https://www.qubes-os.org/doc/issue-tracking/#milestones
> 
> If the documentation is unclear, or you disagree with how things are done and have suggestions for improvement, I'm open to talking about it, but that would be a separate topic of discussion.

Ok.

> > Another example is #5667 - happened during Qubes 4.0. Does it still happen with Qubes R4.1? I don't know. Didn't test yet since that would take time. From my perspective, could even be wasted time. Since there was no reply last time, there might be no reply if I report "R4.1 still affected". But if I don't test and comment, the stale bot would close the ticket.
> 
> No, no one said you have to test _and_ comment. Just a comment is enough, like, "Still need to check on this," "Might affect 4.1," or even just "Please keep open." The bot doesn't know the difference. It just checks for _any activity whatsoever_.

Sounds good!

**Comment by adrelanos** [comment]
Responded: Nothing coming to mind.

What I had in mind was https://github.com/QubesOS/qubes-issues/issues/6585. Root cause is an upstream issue.  And for almost a year it stalled it's a quote [crazy situation](https://github.com/QubesOS/qubes-posts/pull/79#issuecomment-1044051919). But seems like https://github.com/QubesOS/qubes-posts/pull/79 and https://github.com/QubesOS/qubes-issues/issues/6585 can be gotten back to progress by commenting again.

Never mind my concerns. Thank you for your consideration! Thank you for hearing me. Since a stale bot cannot be stopped, is desired, my plan is to see how it goes and provide constructive feedback as it happens.

**Comment by andrewdavidwong** [comment]
> Responded: Nothing coming to mind.
> 

Sorry, I'm having trouble parsing this.

> What I had in mind was #6585. Root cause is an upstream issue. And for almost a year it stalled it's a quote [crazy situation](https://github.com/QubesOS/qubes-posts/pull/79#issuecomment-1044051919). But seems like [QubesOS/qubes-posts#79](https://github.com/QubesOS/qubes-posts/pull/79) and #6585 can be gotten back to progress by commenting again.
> 

Ok, but that situation seems to belong to a different category than the scenario you mentioned: 'The stale bot might close too many issues and when this gets reported as a bug it might be closed as "not our bug".'

IMO, the way to handle the latter would be through normal human communication:

>> Hey, the stale bot is closing issues that it shouldn't be. I'm reporting this as a bug in the stale bot.

>> *Bug report gets closed as "not our bug."*

>> Hey, this got closed as "not our bug," but I think it really is your bug, because you said the bot isn't supposed to close issues with any activity in the past year. I commented on this issue last week, yet the bot still closed it yesterday.

>> *Bug report gets reopened and fixed.*

It sounded like you were assuming that we wouldn't be amenable to such communication, which is perplexing, because I believe we have a good track record of being reasonable about this stuff.

-----

> Never mind my concerns. Thank you for your consideration! Thank you for hearing me. Since a stale bot cannot be stopped, is desired, my plan is to see how it goes and provide constructive feedback as it happens.

This "cannot be stopped" is another example. (Maybe it's just a language barrier or translation issue?) This makes it sound like we've already stubbornly made up our minds and refuse to listen to any reasonable objections or dissenting opinions. I assure you, we do listen to reasonable arguments and objections, as evidenced by the comments on this issue. However, we may not always *agree* with others' conclusions. (Even this use of "we" may be misleading, as we sometimes disagree with each other.)

There is a difference between listening and agreeing. It's tempting to think "anyone who listens to my argument will agree with it" or "if you don't agree, then you weren't listening," but this is simply assuming one's own infallibility, which is out-of-scope in rational, constructive human interaction. :) Ultimately, after listening, whoever is in charge has to make a decision and take action, which means taking on the responsibility of sometimes being wrong. That's just the cost of doing business. The alternative is eternal analysis paralysis.

In the case of this particular issue, nothing is set in stone. The bot hasn't (AFAIK) been created or deployed yet. So far, there's just been commentary and discussion. I respect your opinion, and I respect that you had some reservations and objections. I wanted to do you the courtesy of responding to them honestly and forthrightly.

**Comment by adrelanos** [comment]
> > Responded: Nothing coming to mind.
> 
> Sorry, I'm having trouble parsing this.

"Why would we respond that way? Have we ever done anything like this in the past?"

-> No, nothing comes to mind where written response was like that.

> Ok, but that situation seems to belong to a different category than the scenario you mentioned: 'The stale bot might close too many issues and when this gets reported as a bug it might be closed as "not our bug".'

Ok.

> IMO, the way to handle the latter would be through normal human communication:
> 
> > > Hey, the stale bot is closing issues that it shouldn't be. I'm reporting this as a bug in the stale bot.
> 
> > > _Bug report gets closed as "not our bug."_
> 
> > > Hey, this got closed as "not our bug," but I think it really is your bug, because you said the bot isn't supposed to close issues with any activity in the past year. I commented on this issue last week, yet the bot still closed it yesterday.
> 
> > > _Bug report gets reopened and fixed._

Alright.

> It sounded like you were assuming that we wouldn't be amenable to such communication, which is perplexing, because I believe we have a good track record of being reasonable about this stuff.

Yes.

> > Never mind my concerns. Thank you for your consideration! Thank you for hearing me. Since a stale bot cannot be stopped, is desired, my plan is to see how it goes and provide constructive feedback as it happens.
> 
> This "cannot be stopped" is another example. (Maybe it's just a language barrier or translation issue?)

Yes, possibly text communication issue. No worries. Not a big deal.

> Ultimately, after listening, whoever is in charge has to make a decision and take action, which means taking on the responsibility of sometimes being wrong. That's just the cost of doing business. The alternative is eternal analysis paralysis.

For sure. Don't let the perfect be the enemy of the good. :)

> In the case of this particular issue, nothing is set in stone. The bot hasn't (AFAIK) been created or deployed yet. So far, there's just been commentary and discussion.

Ok.

> I respect your opinion, and I respect that you had some reservations and objections. I wanted to do you the courtesy of responding to them honestly and forthrightly.

Thank you. :)

**Comment by adrelanos** [comment]
This seems messy. ~ 30 tickets just got closed on 12/7/24 where I got 60 e-mail notifications for.
(Default closure message + closing ticket message.)

I am just ignoring this. Replying to ~ 30 tickets, I would almost need automation. And doing that every time stalebot runs, is too tedious. So these tickets will just stay closed, maybe resulting in more duplicates.

Maybe I'll ask to re-open one or another ticket if it is particularly high priority and ignore the test. Maybe that's even the desired result here, having only tickets open which people care about most.

**Comment by marmarek** [comment]
I understand the batch of emails may be annoying... but we don't do that often, qubes release EOL is an event every two years approximately. Just to repeat - we are not closing issues just because they are inactive (I know some projects do, but we consider it a bad idea). We are closing them because they are marked as affecting only a version that is not supported anymore.

> Replying to ~ 30 tickets, I would almost need automation.

It is _not_ about replaying just because ticket was closed. It's about telling if the issue still affecting the newer version too. In fact, I would expect such info posted when user reporting the issue (or someone else affected by it) upgraded to the newer version, when they see it's still there - which would prevent closing such issue in the first place.

And additionally, we intentionally don't close them on the day of end of support, but few months later - to give even more time for users to say if given issue still affect them on the new release (before it gets closed).

If nobody complain about an issue, then perhaps it isn't important issue after all, and we should focus on things that are actually annoying to the users (we surely are not lacking things to do).

**Comment by ben-grande** [comment]
> > Replying to ~ 30 tickets, I would almost need automation.
> 
> It is _not_ about replaying just because ticket was closed. It's about telling if the issue still affecting the newer version too.

I agree with this. I am closing a lot of issues recently, focusing on creation date. A lot of issues I encounter are

- Duplicate:
  - Filtering 2100 issues can give plenty results unless one learns github search syntax. Users don't search. Triage is time consuming.
- Solved:
  - PR didn't add "Fixes"
  - PR did not link to the issue
  - Developer didn't know issue existed and fixed it
- Not planned:
  - Upstream issue
  - User workflow specific, not applicable widely
  - Not a good idea
  - Solved but in another way
  - Unimportant, nobody else complained/asked, not security relevant
  - a lot of other reasons

Stale bot helps us by closing these issues if it is not affecting a release. If it didn't act, would be even more issues to triage. I lot of things I can close in few minutes, some issues takes more time to triage, evaluate, test if it is fixed and reply. But what would absolutely help is users reporting "affects X release", so less issues for us to triage.
